File: platforms/windows/Dimmy.Windows.Tests/Services/UpdateGateTests.cs
using Dimmy.Windows.Services;
using Xunit;

namespace Dimmy.Windows.Tests.Services;

public class UpdateGateTests
{
    [Fact]
    public void An_idle_app_may_be_replaced()
    {
        Assert.True(UpdateGate.MayEndProcess(0));
    }

    [Fact]
    public void A_recording_meeting_blocks_the_update()
    {
        Assert.False(UpdateGate.MayEndProcess(1));
    }

    [Fact]
    public void A_lock_failure_counts_as_recording()
    {
        // We could not read the meeting state. The safe answer is the one
        // that cannot destroy a recording: an update waits, a lost meeting
        // does not come back.
        Assert.False(UpdateGate.MayEndProcess(-1));
    }

    [Fact]
    public void Any_unexpected_return_counts_as_recording()
    {
        // Only an explicit 0 means "nothing is recording". Everything else
        // — a future rc, a garbled value — errs toward keeping the process
        // alive.
        Assert.False(UpdateGate.MayEndProcess(2));
        Assert.False(UpdateGate.MayEndProcess(int.MinValue));
    }
}

File: platforms/macos/DimmyTests/UpdateGateTests.swift
import XCTest

@testable import Dimmy

// MARK: - UpdateGateTests
//
// Pins the pure `UpdateService.mayEndProcess(_:)` decision that vetoes a
// Sparkle check (and a Sparkle install) while a meeting records.
// Installing relaunches the app, which kills the capture threads
// mid-write: the audio survives, but the recording is cut short, never
// finalized, and orphaned with no recap. Burned on Windows 2026-09-04 —
// Velopack applied a staged update 6 minutes into a real meeting. Mac
// parity with `Dimmy.Windows/Services/UpdateGate.cs`.

final class UpdateGateTests: XCTestCase {

    func testAnIdleAppMayBeReplaced() {
        XCTAssertTrue(UpdateService.mayEndProcess(0))
    }

    func testARecordingMeetingBlocksTheUpdate() {
        XCTAssertFalse(UpdateService.mayEndProcess(1))
    }

    func testALockFailureCountsAsRecording() {
        // We could not read the meeting state. The safe answer is the one
        // that cannot destroy a recording: an update waits, a lost meeting
        // does not come back.
        XCTAssertFalse(UpdateService.mayEndProcess(-1))
    }

    func testAnyUnexpectedReturnCountsAsRecording() {
        XCTAssertFalse(UpdateService.mayEndProcess(2))
        XCTAssertFalse(UpdateService.mayEndProcess(Int32.min))
    }
}
